server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -779,10 +779,26 @@ public void relocated(
         final String targetAllocationId,
         final BiConsumer<ReplicationTracker.PrimaryContext, ActionListener<Void>> consumer,
         final ActionListener<Void> listener
+    ) throws IllegalIndexShardStateException, IllegalStateException {
+        relocated(targetNodeId, targetAllocationId, consumer, listener, null);
+    }
+
+    /**
+     * Provides an variant of {@link IndexShard#relocated(String, String, BiConsumer, ActionListener, Releasable)} with an option
+     * to relocate the shard under externally acquired primary permits.
+     *
+     * @param acquiredPrimaryPermits if null, waits until all the primary permits are acquired, otherwise completes relocation immediately
+     */
+    public void relocated(
+        final String targetNodeId,
+        final String targetAllocationId,
+        final BiConsumer<ReplicationTracker.PrimaryContext, ActionListener<Void>> consumer,
+        final ActionListener<Void> listener,
+        @Nullable final Releasable acquiredPrimaryPermits
```

Comment:
I'm a bit concerned about this API change, nothing prevents calling this method with an empty releasable that could cause correctness issues, wdyt?

Reply:
We can add assertion that if it's null, the primary permits are still blocked/held.

Reply:
I've added such an assertion